### PR TITLE
[stream] added rate limiter

### PR DESCRIPTION
**Pull request by JackyWYX**
## Description

This is a small break down PR for https://github.com/harmony-one/harmony/pull/3535. This PR sits on top of https://github.com/harmony-one/harmony/pull/3554

1. Added a ratelimiter module to limit the rate of stream requests.
2. The ratelimiter is limiting: 1. global requests per second 2. request per second for each stream.
3. In upper function call, global rate is set at 50/s, single stream rate is set at 10/s

### Update 03/01/2021

Added stream removed handling according to comment. Now this PR is on top of [stream manager](https://github.com/harmony-one/harmony/pull/3554)

### CODE REVIEW

File: p2p/stream/common/ratelimiter/ratelimiter.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+		globalLimiter: ratelimit.New(global),
+
+		streamRate: single,
+		limiters:   make(map[sttypes.StreamID]ratelimit.Limiter),
+	}
+}
+
+func (rl *rateLimiter) LimitRequest(stid sttypes.StreamID) {
+	rl.lock.RLock()
+	limiter, ok := rl.limiters[stid]
```
Comment (LeoHChen) [inline p2p/stream/common/ratelimiter/ratelimiter.go:43]: limiter is per stream id, will you release the limiter from the map?

**Review by LeoHChen — COMMENTED** [review]
please consider to add prometheus metrics support

File: p2p/stream/common/ratelimiter/ratelimiter.go (hunk 1 above)
Comment (JackyWYX) [reply p2p/stream/common/ratelimiter/ratelimiter.go:43]: Good point. Let me do it.

Comment (JackyWYX) [reply p2p/stream/common/ratelimiter/ratelimiter.go:43]: After looking into the prometheus for a while, I think I will add prometheus support after the merge of PRs for the following reason:
1. The prometheus metrics will not effect the key feature. It's a better-to-have feature.
2. Prometheus metrics can be added in a separate PR which will make the changes more clean.
So I created [an issue](https://github.com/harmony-one/harmony/issues/3555) to be executed after the review of this entire merge.

**Comment by JackyWYX** [comment]
> please consider to add prometheus metrics support

After looking into the prometheus for a while, I think I will add prometheus support after the merge of PRs for the following reason:

The prometheus metrics will not effect the key feature. It's a better-to-have feature.
Prometheus metrics can be added in a separate PR which will make the changes more clean.
So I created an issue to be executed after the review of this entire merge.

**Review by rlan35 — APPROVED** [review]

File: p2p/stream/common/ratelimiter/ratelimiter.go (hunk 1 above)
Comment (JackyWYX) [reply p2p/stream/common/ratelimiter/ratelimiter.go:82]: Resolved. Please review again.

**Comment by JackyWYX** [comment]
@rlan35 @LeoHChen  Added removing stream handling to avoid possible memory leak. Please review again.

**Review by rlan35 — APPROVED** [review]

**Pull request merged**